### PR TITLE
Fix batch mode behavior

### PR DESCRIPTION
**Pull request by maumueller**
A follow-up on #201. I think the batch mode file handling works nicely again (except #203). 

However, the whole docker parallelization rendered batch mode in Docker useless and this PR fixes this. Batch mode was meant to allow an implementation to go 'all in' with CPU/GPU resources. It doesn't help an implementation much if it receives all the queries at once if it then has to solve it in a single thread :-) 

I tried to find a fix that was minimally invasive with the rest of the code base. Let me know what you think.

### CODE REVIEW

**Comment by erikbern** [comment]
I thought the point of batch was for algorithms that use minibatches on the GPU or similar :)

But I don't have any strong feelings about it. However – if the point is to go all in on the GPU/CPU then I wonder if it would make sense to change the default implementation of batch queries to use a thread pool when there's no custom batch query interface?

**Comment by maumueller** [comment]
Nice suggestions, implemented it. Gave a speedup of roughly 3 in QPS for annoy on my 4-core laptop. Noticed another problem in the website creation that I fixed as well.

File: ann_benchmarks/algorithms/base.py [hunk 1]
```diff
@@ -19,9 +19,12 @@ def query(self, q, n):
         return []  # array of candidate indices
 
     def batch_query(self, X, n):
-        self.res = []
-        for q in X:
-            self.res.append(self.query(q, n))
+        """Provide all queries at once and let algorithm figure out
```
Comment (erikbern) [inline ann_benchmarks/algorithms/base.py:22]: isn't there code for

File: ann_benchmarks/algorithms/base.py [hunk 2]
```diff
@@ -19,9 +19,12 @@ def query(self, q, n):
         return []  # array of candidate indices
 
     def batch_query(self, X, n):
-        self.res = []
-        for q in X:
-            self.res.append(self.query(q, n))
+        """Provide all queries at once and let algorithm figure out 
+           how to handle it. Default implementation uses a ThreadPool
+           to parallelize query processing."""
+        from multiprocessing.pool import ThreadPool
```
Comment (erikbern) [inline ann_benchmarks/algorithms/base.py:25]: super minor but i'd have a preference for this at the top module level

**Comment by erikbern** [comment]
Looks great, is this ready to be merged? Happy to do so whenever

File: ann_benchmarks/algorithms/base.py (hunk 2 above)
Comment (maumueller) [reply ann_benchmarks/algorithms/base.py]: Fixed.

**Comment by maumueller** [comment]
Sorry, lost track of this. I squashed the commits and rebased. Ready to be merged. 

(I was about to sneak in adding an explicit pointer to the interactive visualisations from the plots in `README.md`, but maybe this should be its own PR. :-))

**Comment by erikbern** [comment]
sorry dropped the ball on this (along with some other ann-benchmarks stuff), but merged it!

**Pull request merged**